### PR TITLE
Add memory control for MergeReader

### PR DESCRIPTION
**Pull request by lancelly**
In this pr, I extract interface MemoryReservationManager. MemoryReservationManager is used to reserve/release memory from LocalExecutionPlanner#freeMemoryForOperators during planning and execution stage. This pr adds memory control logic of PriorityMergeReader.
One FI may have multiple SeriesScanUtils, each FI only has one MemoryReservationManager and the SeriesScanUtils use MemoryReservationManagerWrapper to queue the reserve/release operations.
The most important thing is to make sure reservedBytesIntal of MemoryReservationManager is correctly returned to freeMemoryForOperators.

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/NotSynchronizedMemoryReservationManager.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.queryengine.plan.planner.memory;
+
+import org.apache.iotdb.db.queryengine.common.QueryId;
+import org.apache.iotdb.db.queryengine.plan.planner.LocalExecutionPlanner;
+
+public class NotSynchronizedMemoryReservationManager implements MemoryReservationManager {
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/NotSynchronizedMemoryReservationManager.java:25]:
```suggestion
public class NotThreadSafeMemoryReservationManager implements MemoryReservationManager {
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/SynchronizedMemoryReservationManager.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.queryengine.plan.planner.memory;
+
+import org.apache.iotdb.db.queryengine.common.QueryId;
+
+public class SynchronizedMemoryReservationManager extends NotSynchronizedMemoryReservationManager {
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/SynchronizedMemoryReservationManager.java:24]:
```suggestion
public class ThreadSafeMemoryReservationManager extends NotThreadSafeMemoryReservationManager {
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/SynchronizedMemoryReservationManagerWrapper.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * SynchronizedMemoryReservationManager synchronizes the memory reservation and release operations.
+ * However, synchronization is not necessary every time for the memory reservation and release
+ * operations in single thread context. For example, each SeriesScanUtil itself reserves and
+ * releases memory in a single thread context, but SeriesScanUtils in the same FI may share the same
+ * SynchronizedMemoryReservationManager. So we can use this wrapper to batch the reserve and release
+ * operations to reduce the overhead of synchronization. The actual reserve and release operations
+ * are delegated to the wrapped MemoryReservationManager.
+ */
+public class SynchronizedMemoryReservationManagerWrapper implements MemoryReservationManager {
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/SynchronizedMemoryReservationManagerWrapper.java:31]: do we really need this?

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
change the memory estimate * 3 in ScanOperator

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/NotSynchronizedMemoryReservationManager.java (hunk 1 above)
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/NotSynchronizedMemoryReservationManager.java]: Done.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/SynchronizedMemoryReservationManager.java (hunk 2 above)
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/memory/SynchronizedMemoryReservationManager.java]: Done.

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**